File: Src/Runtime/Strings/Trim.cpp
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/

//
// Created by Rodrigo on 5/24/26.
//

#include "ADT/List/Object.h"
#include "Trim.h"

#include "TrimLeft.h"
#include "TrimRight.h"

using namespace Typed;
using namespace Typed::Runtime;
using namespace Typed::Runtime::Strings;

ADT::Runtime::Object Strings::Trim(
    ADT::List::Object &args,
    ADT::Lang::AST *_
)
{
    auto left = TrimLeft(args, _);
    args[0] = std::move(left);

    return TrimRight(args, _);
}

File: src/runtime/strings/trim.rs
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/
use crate::adt::lang::{ASTValue, RuntimeArguments, AST};
use crate::adt::result::ExecutionTupleResult;
use crate::adt::runtime::{HashableObject, Object};
use crate::support::runtime::execution::continue_execution;
use crate::support::runtime::object::get_string;

pub fn trim<'a>(
    args: RuntimeArguments<'a>,
    trace: &AST<'a>
)
    -> ExecutionTupleResult<'a>
{
    let origin_obj = args.get(0).unwrap();
    let origin = get_string(origin_obj, trace)?;
    
    // If the string is static, we can trim it without allocating
    if let Object::Hashable(
        HashableObject::String(
            ASTValue::Borrowed(s)
        )
    ) = origin_obj {
        let trimmed = s.trim();
        return continue_execution(
            Object::Hashable(
                HashableObject::String(
                    ASTValue::Borrowed(trimmed)
                )
            )
        );
    }
    
    // Otherwise, we need to allocate a new string
    let trimmed = String::from(origin.trim());
    
    continue_execution(
        Object::Hashable(
            HashableObject::String(
                ASTValue::Owned(trimmed)
            )
        )
    )
}